### ISSUE TITLE
Add a way to map multichannel buses using argument-string syntax

### ISSUE THREAD

**Issue by jamshark70** [supercollider/supercollider#336]
Currently there are two ways to map single-channel buses onto SynthDef arguments:
- OSC message (/n_map, /n_mapn)
- Argument syntax: `"argname", "c/n/"` where /n/ is the bus number

But there is only one way to map a multichannel bus onto an array argument:
- /n_mapn

I propose extending argument syntax: `c/n/_/c/` where /c/ is the number of channels to map, e.g. `\freqs, \c0_4` to map four consecutive control buses onto four consecutive inputs. If not given, 1 would be assumed (current behavior).

The change on the language side would be trivial. It would also require a change in the server to handle the number of channels. I believe this wouldn't be especially hard but I don't know the server code very well.

Possible drawback: Currently, you can accidentally map a multichannel bus onto a non-array control, and the additional bus channels do not spill over onto the following controls. That would change. But I think it's reasonable that the user should be responsible for managing the number of bus channels properly.

**Comment by sonoro1234** [comment]
I think that /n_mapn already works as this. ("argname", busnum, number of consecutive busses)
Should issue be closed?

**Comment by telephon** [comment]
Good idea.

>Currently, you can accidentally map a multichannel bus onto a non-array control, and the additional bus channels do not spill over onto the following controls. That would change. 

I don't think it would change, because when you use a control name, the server side checks how many channels a control has. It shouldn't spill, also ought not to in the future.

**Comment by Sciss** [comment]
I don't understand the proposal. As @sonoro1234 noted, `/n_mapn` already works correctly for multi-channel arguments.

**Comment by jamshark70** [comment]
The point is to improve user friendliness.

Where should a user get any sort of clue to expect that `Bus:asMap` works for single-channel buses, but not for multichannel ones?

- Documentation: Currently this limitation is undocumented.

- Warnings: Nope, not there either. Our code just blindly assumes it will be fine, while only mapping the first channel (i.e. failing silently).

So there's basically one way left to figure it out: try it, get a confusing result, and waste creative time trying to debug it. This makes users unhappy.

So the proposal is to make `c...` mapping strings work compatibly for single- and multichannel buses -- that is, to provide a fully working feature, instead of a half-working one. One rationale is that you can use mapping strings easily in patterns, but mapn is a right pain in the **** for that.

If that's too difficult, I would be ok with documenting the limitation and warning the user about it in code. But I think that's a weak fix. If possible, we should make things work the way users will intuitively expect.

**Comment by sonoro1234** [comment]
I see, your proposal is not a change in scsynth protocol (/n_mapn) but for sclang being more friendly with it.

**Comment by Sciss** [comment]
@jamshark70 perhaps I just don't understand what is meant by

>It would also require a change in the server to handle the number of channels.

Can't you just improve the pattern mapping syntax and always use /n_mapn? In my system SoundProcesses AFAIK I have simply given up using /n_map and always use /n_mapn as it's more universal.

**Comment by jamshark70** [comment]
> I see, your proposal is not a change in scsynth protocol (/n_mapn) but for sclang being more friendly with it.

The proposal is to extend `"/n_set", id, "name", "c1"` to support multiple channels.

The catch is, `"c1"` by itself does not carry any information about the number of channels.

If we want to include that, then the string format needs to be extended, e.g. `"ci_n"` where `i` is the bus index and `n` is the number of channels (`c1_2`).

Currently, scsynth will not understand that format. So, it does involve scsynth.

> Can't you just improve the pattern mapping syntax and always use /n_mapn?

No, I don't think that will be enough.

People are already writing things like

```
p = Pbind(
	... blahblah...,
	\pan, panBus.asMap
).play
```

From the user's perspective, "why does this work for single-channel buses but not multichannel?" is a valid usage question. So far, nobody has provided any convincing reason why it should be this way.

In the end, it might turn out that the best solution is documentation (perhaps, consider it a corner case that doesn't justify the development time). But I'm a little unsatisfied at the moment because I think the user's perspective here is not being fully understood.

**Comment by Sciss** [comment]
My position is always conservative with respect to server message change. From what I understand, this can be totally solved in the language. Both `/n_set` and `/n_setn` already support multi-channel arguments:

> **/n_set**
> This message now supports array type tags ($[ and $]) in the control/value component of the OSC message. Arrayed control values are applied in the manner of n_setn (i.e., sequentially starting at the indexed or named control).

I think to convince me, you need to sketch out the entire scenario, and why what you are trying to do is not possible with the current api. You can do

    "/n_set", id, "name", [c1_0, c1_1, ...]

And if you put an array here is doesn't matter if c1 is single or multi channels. The same goes for `/n_setn`. Perhaps I'm just not seeing the obstacle for your use case here?

**Comment by Sciss** [comment]
So to summarise my point, your problem seems to be that `Pbind` or event-stream-player doesn't correctly splice multi-channel-bus parameters, and thus that's the place to fix the problem, because the server API supports that case.

**Comment by jamshark70** [comment]
By that logic, then, `c0` syntax was a mistake in the first place, because it does something that you can already do with `/n_map`.

But now we have a half-complete implementation. We can leave it as is (internally inconsistent), complete it or remove it, then.

**Comment by telephon** [comment]
@jamshark70 if you look at `BusPlug`, it does this:

```supercollider
makeBusArg {
		var index, numChannels, prefix;
		if(bus.isNil) { ^busArg = "" }; // still neutral
		prefix = if(this.rate == \audio) { "\a" } { "\c" };
		index = this.index;
		numChannels = this.numChannels;
		^busArg = if(numChannels == 1) {
			prefix ++ index
		} {
			{ |i| prefix ++ (index + i) }.dup(numChannels)
		}
	}
```

so when you embed it in an event, it will expand in such a way. My decision was that it should multichannel expand the event. So a Bus should perhaps expand in a `Ref` of an array.

**Comment by Sciss** [comment]
(Sorry, I simply have no idea was "`c0` syntax" is, so just reiterating, in order for the issue ticket to be more understandable and accessible, it needs to be clarified what exactly a piece of code is that doesn't work, and how it should work. This is all very fragmentary. Julian's last reply suggests that indeed it is a question of fixing a class library method.)

**Comment by telephon** [comment]
@Sciss could we please discuss this without haste? To me, James' suggestion is quite reasonable, and it wasn't the purpose of my example to immediately end all consideration.

**Comment by telephon** [comment]
The "`c0` syntax" is a well established way to specify mappings in a `set` method. James McCartney had added it. It makes direct OSC messaging more straightforward.

**Comment by Sciss** [comment]
No haste, just trying to understand _what_ is happening :)

**Comment by sonoro1234** [comment]
I had no idea of the
```
"/n_set",
```

**Comment by jamshark70** [comment]
Ok, history.

In early SC Server, the only way to initialize a synth node with one or more controls mapped to a bus was to send a bundle with `/s_new` and `/n_map`. After awhile, users felt this was clumsy for a number of reasons:

- Different code structure for mapped vs non-mapped synths. That is, if you write your code first without mappings, and then you need mapping later, you have to add `sendBundle` and wrap the synth code in a function.

- Difficult to integrate into patterns. Yes, it's possible to have the event prototype generate the pair of messages, but that will certainly be slower than what we have now (and granular synthesis with patterns is a use case we shouldn't rule out by impairing performance).

So, with James McCartney's addition, you can write:

```
Synth(\demo, [freq: 220, pan: panBus.asMap])
```

... and the message will include `"freq", 220, "pan", "c3"` (or another bus index), and scsynth interprets that to map pan to kr bus 3. One message, integrates into Synth.new and patterns easily, no fuss.

But it only works for single channel buses.

Either the initial rationale was correct, or it wasn't. If it was correct, we should extend it to cover all buses. If it wasn't correct, then we should deprecate -- but I'm fairly sure this is a widely popular feature. I was just looking at TX Modular's code base today -- `c` arguments are everywhere in it. We really can't get rid of it now.

So there are really only two options: document it as an annoying corner case, or fix it.

**Comment by jamshark70** [comment]
> This is only necessary to use /n_set for doing /n_map and /n_mapa work...

No, it is also valuable in s_new. It is even more commonly used in s_new!

**Comment by scztt** [comment]
I would second @sonoro1234 w/r/t not changing the server API. The `n_set` / `n_map` and `set_n`/`n_mapn` variants both support multichannel buses - via grouping e.g. `[ 'n_set', 1004, "in", "[", "a6", "a7", "]"` or via bus counts in the `_n` case, `[ 'n_setn', 1004, "in", 2, "a6", "a7" ]`. If we wanted to be strict, we could deprecate one of these, since I believe the functionality is entirely overlapping - but I don't see good reason to change anything.

The problem seems like an sclang-side problem. James is being too kind by calling this a "user friendliness" issue - right now, `Bus.audio(s, 2).asMap` return `a4` which is just plain wrong, since it effectively loses the channel count. In JITLib, we get `Ndef(\foo, { DC.ar([0,0]) }).asMap == [a4, a5]` - this should be the correct behavior for the `Bus` object as well. The workaround when using .set is already disgusting: `node.set(bus.numChannels.collect(bus.subBus(_)).collect(_.asMap))`. The workarounds required when using the Event / Pattern system are much more subtle and tricky (I spent the whole weekend fiddling with these sorts of problems, unfortunately).

@telephon Was the protection against overwriting adjacent arguments added recently (e.g. last few years)? I have distinct (and traumatic) memories of 8 hour days spent tracking down bugs related to overflowing arg arrays.

**Comment by telephon** [comment]
>Was the protection against overwriting adjacent arguments added recently (e.g. last few years)? 
Yes, it's bee a while, but not forever. I am not sure we have tests for it though.

**Comment by telephon** [comment]
>this should be the correct behavior for the Bus object as well. 

no, it would multichannel expand the event, like NodeProxy does.

It might be possible that `Bus.asMap` could return something like `Ref(["a0", "a1"])` for a stereo audio bus. The `Ref` is necessary for avoiding multichannel expansion.

**Comment by scztt** [comment]
Hmm - why wrap the result of `.asMap` in a Ref, but not the result of any other array-returning methods? The normal pattern is to wrap Event / Pbind values to prevent multichannel expansion - it feels weird to make `.asMap` as special case, where you'd need to *unwrap* it to achieve multichannel expansion, rather than wrap to avoid. The event case is the only one I'm aware of where expansion would come into play.

And, additionally: `Ref` is *not* the correct way to pass arrayed arguments (as I recently learned). Ref's behave like Functions - depending on the Event type, Event key, and code path being hit, Ref's may either be wrapped or left unwrapped in unexpected ways, because of Event's tendency to access properties via e.g. `~prop.value`. Arrayed arguments are supposed to be wrapped in an additional array, e.g. `[[\a1, \a2, \a3]]` (see http://doc.sccode.org/Classes/Event.html#Arrayed%20Arguments)

An example of weird Ref functionality with Events: 
```
(instrument: \threeF, degree: `[1, 2, 3]).play;
->  [ 9, "threeF", 1004, 0, 1, "out", 0, "freq", DATA[16] ]
```
```
(instrument: \threeF, freq: `[1, 2, 3]).play;
->  [ "#bundle", 16033062952253253014, 
  [ 9, "threeF", 1004, 0, 1, "out", 0, "freq", 100 ],
  [ 9, "threeF", 1005, 0, 1, "out", 0, "freq", 200 ],
  [ 9, "threeF", 1006, 0, 1, "out", 0, "freq", 500 ]
]
```

**Comment by telephon** [comment]
>: Ref is not the correct way to pass arrayed arguments (as I recently learned).
OK, yes, that is a point. We need a more systematic handling of placeholders, as it seems again and again. So good, return an array of arrays.


>Hmm - why wrap the result of .asMap in a Ref, but not the result of any other array-returning methods? 

There are two cases we need to distinguish:
we can map multiple channels of a single bus object to:

1. each single control channels of multiple synths 
2. multiple control channels of a single synth

We are currently talking about (2).

**Comment by jamshark70** [comment]
> I would second @sonoro1234 w/r/t not changing the server API.

I was concerned about performance -- event type \note is slow enough as it is -- but...

> The n_set / n_map and set_n/n_mapn variants both support multichannel buses - via grouping e.g. [ 'n_set', 1004, "in", "[", "a6", "a7", "]"...

Hm, that just might do it without hurting performance too much.

> If we wanted to be strict, we could deprecate one of these, since I believe the functionality is entirely overlapping

A key advantage of mapping strings is that you can use them in s_new -- but for some reason, multiple posters have been talking about n_map vs n_set only. I don't think they do overlap totally: n_map can't do the same as s_new.

n_map was there first -- deprecating would be risky.

"c0" syntax is, by now, indispensable.

I kind of like the array-arg approach, if we can make it transparent to the user. Maybe a bus-mapping object that responds to asControlInput appropriately.

**Comment by telephon** [comment]
In `Event`, we call `msg.asOSCArgArray`. (https://github.com/supercollider/supercollider/blob/develop/SCClassLibrary/Common/Collections/Event.sc#L550). This is where any bus object could be translated into a suitable shape. Note that the multichannel expansion happens before that (in `flopTogether`).

```
Bus {
     asOSCArgArray {
         ^this.mapArgs.asOSCArgArray // assuming that mapArgs can be something like ["a0", "a1"]
     }
}
```

might do it.

Synths don't multichannel expand anyhow.

**Comment by jamshark70** [comment]
> In Event, we call msg.asOSCArgArray. (https://github.com/supercollider/supercollider/blob/develop/SCClassLibrary/Common/Collections/Event.sc#L550). This is where any bus object could be translated into a suitable shape.

You don't know whether the synth expects a bus index or bus mapping. If you change Bus to return a mapping string by default, it will break e.g. `out: aBusObject` in an event.

While I'm often reluctant to add a new object, here it would be conceptually clear: the event would contain a single object whose name explains unequivocally that it will be a bus mapping, and it would expand to the array in OSC messages.

I'm not crazy about using Ref for this. I think it'll make debugging harder.

**Comment by telephon** [comment]
(yes, see my edit above)

>I'm not crazy about using Ref for this. I think it'll make debugging harder.

it won't work, as @scztt mentioned above, because `Ref` answers its content on `.value`. We need a dedicated simple wrapper. Its name should mirror its semantics as good as we can.

**Comment by scztt** [comment]
@telephon - we have to be careful with calling asMap implicitly. Right now (in most common cases I'm aware of), passing a bus as an synth argument turns it in to the index of that bus - we should not be calling `.asMap` implicitly anywhere, else we would disrupt this behavior. `node.map` forms already take an bus index (not a `/c4`), and for `node.set` / `Synth(args:[])` forms, mapping should be considered non-default and should require an `asMap` call outside the API.

**Comment by scztt** [comment]
Event cases are the hardest by far, because we have two separate kinds arrayed values we need to expand at different stages. One is multi-event expansion, one is arrayed argument expansion. Actually, you *might* want to say there are three - Patterns often take arrays, sometimes nested ones. We can say that, by default, patterns consume all unwrapped arrays so we don't need additional annotation for those, but maybe it's helpful to call it out explicitly anyway.

So, in an ideal world (where code expressiveness isn't an issue...), we would have something like:
```
SynthDef(\additive, { Out.ar(0, SinOsc.ar(\freq.kr(0 ! 3))) });
Pbind(
    \freq, Pseq([
          Chord([
                Param([100, 200, 300]), Param([130, 260, 390])
          ]),
          Chord([
                Param([150, 300, 450]), Param([200, 400, 600])
          ])
    ])
)
```
This would result in chord-pairs of events that looked like `(freq: [100, 200, 300])`. Right now, the above sort of works for SOME event keys (I don't think it works with \freq, but that's a very special case) - it's less verbose than this example, but much more confusing / hard to understand and write:
```
SynthDef(\additive, { Out.ar(0, SinOsc.ar(\freq.kr(0 ! 3))) });
Pbind(
    \freq, Pseq([
          [
                [[100, 200, 300]], [[130, 260, 390]]
          ],
          [
                [[150, 300, 450]], [[200, 400, 600]]
          ]
    ])
)
```
WTF? That's a lot of brackets. My first example seems (to me) to have mostly unambiguous ways to expand. During pattern parsing, we stop when we hit a `Chord` or `Param` (most patterns only go 1 level deep, so this is a bit moot - but there are a few patterns that go 2 deep, which makes the second example even weirder). During multichannel expansion, we iterate our Event values and expand out every `Chord[]` array. During argument generation, we turn cases of `Param` into array arguments. I think the rough rule is -- Chord and Param behave like arrays when op'd with other arrays/values (so you can do `100 * Chord[1, 2, 3]`, but cannot be op'd together (`Chord[1, 2, 3] * Param[100, 200, 400]` is too ambiguous). 

Since we use `.value` for unboxing throughout Event, we CAN'T use `.value` to unbox these - it has to be something else. I would guess this might look something like:
```
+Object { ch { ^ChordArray(this) }    pr { ^ParamArray(this) } }
AbstractBoxedObject { 
    var obj;
    boxType { this.subclassResponsibility(\boxType) } 
    unbox { |type| if (type == this.boxType) { ^obj } { ^this } }
}
ChordArray : AbstractBoxedObject { boxType { ^\chord } }
ParamArray : AbstractBoxedObject { boxType { ^\param } }
```
We might imagine an `AbstractBoxedObject` that derives from Ref, except that it's `.value` method just returns `this` (e.g. no unboxing), and it has another unambiguous way to unbox. Interestingly, I feel like this dovetails with other discussions around wrapping `Rest`.

I think if we're not considering the Pbind/Event ecosystem, the answer is a whole lot easier: `bus.asMap` needs to return an array like `[\c1, \c2, \c3]`. If it does this, everything will work as expected for `set`, `setn`, and `Synth(arg:[])`.

**Comment by telephon** [comment]
thanks for the good discussion. I think the event ecosystem unboxes after multichannel expansion, a slight disadvantage which turns out as advantage here. My suggestion above should take care of it if we have a simple wrapper like the one you describe.

Another reason why I'd like to have an `AbstractObject` superclass, but well, for the future!

**Comment by JordanHendersonMusic** [comment]
This conversation seems to have stalled on abstract object, is there any way forward with the current implementation?
